### PR TITLE
Android EXIF reader

### PR DESCRIPTION
**Pull request by sklencar**
A new expression function introduce to read EXIF data. Currently only for Android.

Sample project: `lutraconsulting/test_exif` (should have public access on Mergin)
Supported functions: 

- `read_exif('<ABSOLUTE_PATH_TO_IMAGE>', '<EXIF_TAG_STRING>')`
-  `read_exif_img_direction('<ABSOLUTE_PATH_TO_IMAGE>')`
-  `read_exif_latitude('<ABSOLUTE_PATH_TO_IMAGE>')`
-  `read_exif_longitude('<ABSOLUTE_PATH_TO_IMAGE>')`

In case of invalid path, invalid tag or missing value, empty value is returned.
It calls java function from EXIFUtils and formats the result.

Manual test added to https://github.com/lutraconsulting/input-manual-tests/issues/5 

Screenshot_20210323-091027

### CODE REVIEW

**Comment by sklencar** [comment]
Some remarks:
- read_exif(file, tag) returns raw data as stored in the tag
- read_exif_img_direction(file) returns angle in degrees (as a number, not formatted string)
- read_exif_latitude(file), read_exif_longitude(file) return lat/lon value in degrees (again as a number, not a string)
- when calculating lat/lon from deg/min/sec please use double not float otherwise we can loose precision (float has too low precision!)
- in formatRationalNumberString() does the conversion work correctly? it looks like you are using integer division there

**Comment by sklencar** [comment]
Remarks resolved, updated info + screenshot

**Comment by saberraz** [comment]
Tested with the APKs and none worked for me. They all come as blank.

**Comment by sklencar** [comment]
https://github.com/lutraconsulting/input/pull/1292 fixes applying exif expressions functions immediately after photo is captured.

https://user-images.githubusercontent.com/6735606/113024297-e0e9dc80-9186-11eb-947e-e1b28f5d846e.mp4

**Comment by sklencar** [comment]
Rebased with master (and fix for applying default values on update)

**Comment by sklencar** [comment]
Found one issue after rebasing with master - crash occurs after modifying an image field. Fix on a way.

**Comment by sklencar** [comment]
Added a handling of invalid values

File: app/inputexpressionfunctions.cpp [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "inputexpressionfunctions.h"
+
+InputExpressionFunctions::InputExpressionFunctions()
```
Comment (PeterPetrik) [inline app/inputexpressionfunctions.cpp:12]: = default

File: app/inputexpressionfunctions.h [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+#ifdef ANDROID
+          if ( values.size() != 2 ) return QVariant();
+
+          QString filepath( values.at( 0 ).toString() );
+          QString exifTag( values.at( 1 ).toString() );
+          return AndroidUtils::getExifInfo( filepath, exifTag );
+# else
+          return QString();
+#endif
+        }
```
Comment (PeterPetrik) [inline app/inputexpressionfunctions.h:58]: move impl to cpp

File: app/inputexpressionfunctions.h [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+
+          double resultDouble = InputUtils::convertRationalNumber( resultString );
+          // If invalid value; expecting [0-360]
+          if ( resultDouble == -1 )
+            return QVariant();
+
+          return QVariant( resultDouble );
+# else
+          return QString();
+#endif
```
Comment (PeterPetrik) [inline app/inputexpressionfunctions.h:95]: move impl to cpp

File: app/inputexpressionfunctions.h [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+          QString filepath( values.at( 0 ).toString() );
+          QString resultString = AndroidUtils::getExifInfo( filepath, GPS_LAT_TAG );
+          if ( resultString.isEmpty() )
+            return QVariant();
+
+          return QVariant( InputUtils::convertCoordinateString( resultString ) );
+# else
+          return QString();
+#endif
+        }
```
Comment (PeterPetrik) [inline app/inputexpressionfunctions.h:129]: move impl to cpp

File: app/inputexpressionfunctions.h [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+          QString filepath( values.at( 0 ).toString() );
+          QString resultString = AndroidUtils::getExifInfo( filepath, GPS_LON_TAG );
+          if ( resultString.isEmpty() )
+            return QVariant();
+
+          return QVariant( InputUtils::convertCoordinateString( resultString ) );
+# else
+          return QString();
+#endif
+        }
```
Comment (PeterPetrik) [inline app/inputexpressionfunctions.h:162]: move impl to cpp

File: app/inputexpressionfunctions.h [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+
+        QgsScopedExpressionFunction *clone() const override { return new ReadExifLongitude();}
+    };
+
+    InputExpressionFunctions();
```
Comment (PeterPetrik) [inline app/inputexpressionfunctions.h:167]: why we need custom constructor, if needed, move to the top of the file also with the register Functions function

File: app/inputexpressionfunctions.h [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+#include "qgsexpressionfunction.h"
+
+#include "androidutils.h"
+#include "inpututils.h"
+
+const static QString GPS_DIRECTION_TAG = "GPSImgDirection";
+const static QString GPS_LON_TAG = "GPSLongitude";
+const static QString GPS_LAT_TAG = "GPSLatitude";
+
+class InputExpressionFunctions
```
Comment (PeterPetrik) [inline app/inputexpressionfunctions.h:26]: add doc

File: app/inpututils.cpp [hunk 8]
```diff
@@ -140,6 +140,31 @@ void InputUtils::setExtentToFeature( const QgsQuickFeatureLayerPair &pair, QgsQu
   mapSettings->setExtent( currentExtent );
 }
 
+double InputUtils::convertCoordinateString( const QString &rationalValue )
```
Comment (PeterPetrik) [inline app/inpututils.cpp:143]: create unit test

File: app/inpututils.cpp [hunk 9]
```diff
@@ -140,6 +140,31 @@ void InputUtils::setExtentToFeature( const QgsQuickFeatureLayerPair &pair, QgsQu
...
+  if ( values.size() != 3 ) return 0;
+
+  double degrees = convertRationalNumber( values.at( 0 ) );
+  double minutes = convertRationalNumber( values.at( 1 ) );
+  double seconds = convertRationalNumber( values.at( 2 ) );
+  double result = degrees + minutes / 60 + seconds / 3600;
+  return result;
+}
+
+double InputUtils::convertRationalNumber( const QString &rationalValue )
```
Comment (PeterPetrik) [inline app/inpututils.cpp:155]: create unit test

File: app/inpututils.cpp [hunk 10]
```diff
@@ -140,6 +140,31 @@ void InputUtils::setExtentToFeature( const QgsQuickFeatureLayerPair &pair, QgsQu
...
+  if ( rationalValue.isEmpty() )
+    return -1;
+
+  QStringList number = rationalValue.split( "/" );
+  if ( number.size() != 2 )
+    return -1;
+
+  double numerator = number.at( 0 ).toDouble();
+  double denominator = number.at( 1 ).toDouble();
+  return numerator / denominator;
```
Comment (PeterPetrik) [inline app/inpututils.cpp:166]: whatif denominator == 0

File: app/inputexpressionfunctions.h [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+          return QString();
+#endif
+        }
+
+        QgsScopedExpressionFunction *clone() const override { return new ReadExifLongitude();}
+    };
+
+    InputExpressionFunctions();
+
+    static void registerFunctions();
```
Comment (PeterPetrik) [inline app/inputexpressionfunctions.h:169]: I am not convinced we need another class `InputExpressionFunctions()`. Cannot be `registerFunctions` just another static function of InputUtils ?

**Comment by sklencar** [comment]
Unit tests were omitted in this PR for now (and created a generic ticket for implementing tests for utils functions).
However, added more checks for util functions used with EXIF functions.

**Comment by PeterPetrik** [comment]
test: https://github.com/lutraconsulting/input-manual-tests/issues/5 -> Test G
docs: TODO

**Comment by sklencar** [comment]
Added docs: https://github.com/lutraconsulting/input-docs/pull/29
Added manual test: https://github.com/lutraconsulting/input-manual-tests/issues/5 (Test H)

**Pull request merged**